src/SimplePie.php

```diff
@@ -1924,6 +1924,7 @@ protected function fetch_data(&$cache)
                             $this->status_code = 0;
 
                             if ($this->force_cache_fallback) {
+                                $this->data['cache_expiration_time'] = $this->cache_duration + time();
                                 $cache->set_data($cacheKey, $this->data, $this->cache_duration);
```

Comment:
If I understand this correctly, this just extends the cache lifetime (`$this->data` remains the same). Not sure why bb38d740c12a953dce4e2722a8d26d7a3ed2be3f made SimplePie store `cache_expiration_time` in the cache when `set_data` already takes TTL but it makes sense to update `cache_expiration_time` here too.

Probably should be a separate commit from the test fix, though.

Reply:
The tests are still related to cache, and were preventing this PR from passing the tests, but can be put in another PR

Reply:
@jtojnar `$this->data['cache_expiration_time']` was introduced to allow mimic the functionality of `mtime()` and `touch()`, because that allowed us to use the cached data even if it was expired. The problem

Reply:
Hmm, maybe we should just pass TTL as `null` (unlimited) to Psr16 caches, and have another layer that would emulate the old `mtime()`/`touch()` API, if we want to have access to expired cache entries. I do not see how can we fix the cache fallback otherwise.

In the long term, we can probably come up with a nicer API though. For example, we could have a class `CacheWrapper` that would handle expiration and invalidation transparently – it could store the cache entry as follows

```php
/** @type CacheEntry array{mtime: int, extraKeys: array<string, mixed>, value: mixed} */
```

and be used something like this:

```php
$cacheWrapper = new CacheWrapper(new Psr16($this->cache), $this->cache_duration);

$this->data = $cacheWrapper->get(
    $key,
    // Extra validation keys, since the allowed size of `$key` is limited.
    // If value of any of the keys does not match the one in the `extraKeys` array,
    // the value will be considered stale and one of the following actions will be performed.
    [
        // Different version of SimplePie might have unrecognizable structure of the cache data
        // so pass `null` as `$staleData` to the callback.
        ['key' => 'build', 'expected' => Misc::get_build(), 'action' => CacheWrapper::INVALID_ACTION_DISCARD],
        // A collision. Cause the value returned by the callback to be returned by `CacheWrapper::get()` but the cache would not be updated.
        // I guess this could also be handled transparently by having the `CacheWrapper` store it in `CacheEntry`.
        ['key' => 'url', 'expected' => $this->feed_url, 'action' => CacheWrapper::INVALID_ACTION_COLLISION],
        // …
    ],
    // This callback will be executed when the cache lacks the key, or it is invalid.
    // `$staleData` will be null if the key did not exist.
    // The cache will be updated with the returned value and the mtime will be updated.
    // Except maybe when exception occurs, then we might want to unset the cache and propagate the exception.
    function(?array $staleData) use (&$obtainedFromCache): array {
        $headers = [];
        if ($staleData !== null && isset($staleData['headers']['etag'])) {
            $headers['etag'] = $staleData['headers']['etag'];
        }

        $file = fetch($url, $headers);

        if ($this->status_code === 304) {
            $obtainedFromCache = true;
            return $staleData;
        }

        // …

        return  [
            'url' => $this->feed_url,
            'feed_url' => $file->get_final_requested_uri(),
            'build' => Misc::get_build(),
            'cache_expiration_time' => $this->cache_duration + time(),
        ];
    }
);
```

Reply:
I would suggest to address this potential refactoring in another PR though :-)
The current PR fixes some relatively obvious bugs, so merging it would already be beneficial

Reply:
P.S. Regarding the tests, the change is due to the fix of the inequality higher up, so they are directly related